packages/lesswrong/lib/collections/ultraFeedEvents/newSchema.ts

```diff
@@ -0,0 +1,171 @@
+import { DEFAULT_CREATED_AT_FIELD, DEFAULT_ID_FIELD } from "@/lib/collections/helpers/sharedFieldConstants";
+
+const ALLOWED_COLLECTION_NAMES = ["Posts", "Comments", "Spotlights"];
+const ALLOWED_EVENT_TYPES = ["served", "viewed", "expanded"];
+
+const schema = {
+  _id: DEFAULT_ID_FIELD,
+  createdAt: DEFAULT_CREATED_AT_FIELD,
+
+  documentId: {
+    database: {
+      type: "TEXT",
+      nullable: false,
+    },
+    graphql: {
+      outputType: "String",
+      inputType: "String!",
+      canRead: ["admins"],
+      canCreate: ["members"],
+      validation: {
+        optional: false,
+      },
+    },
+  } satisfies NewCollectionFieldSpecification<"UltraFeedEvents">,
+
+  collectionName: {
+    database: {
+      type: "TEXT",
+      nullable: false,
+    },
+    graphql: {
+      outputType: "String",
+      inputType: "String!",
+      canRead: ["admins"],
+      canCreate: ["members"],
+      validation: {
+        allowedValues: ALLOWED_COLLECTION_NAMES,
+        optional: false,
+      },
+    },
+  } satisfies NewCollectionFieldSpecification<"UltraFeedEvents">,
+
+  eventType: {
+    database: {
+      type: "TEXT",
+      nullable: false,
+    },
+    graphql: {
+      outputType: "String",
+      inputType: "String!",
+      canRead: ["admins"], // Changed from guests
+      canCreate: ["members"],
+      validation: {
+        allowedValues: ALLOWED_EVENT_TYPES,
+        optional: false,
+      },
+    },
+  } satisfies NewCollectionFieldSpecification<"UltraFeedEvents">,
+
+  userId: {
+    database: {
+      type: "TEXT", // Stores the user ID string
+      nullable: false, // Assuming an event always has a user initiator
+    },
+    graphql: {
+      outputType: "String", // Outputs the ID string
+      inputType: "String!",
+      canRead: ["admins"], // Changed from guests
+      canCreate: ["members"],
+      // Add resolver here later if you want to automatically fetch the User object
+      // resolveAs: { ... }
+      validation: {
+        optional: false,
+      },
+    },
+    form: {
+        label: "User ID",
+        // hidden: true // Often automatically filled or not user-editable
+    }
+  } satisfies NewCollectionFieldSpecification<"UltraFeedEvents">,
+
+  event: {
+    database: {
+      type: "JSONB",
+      nullable: true,
+    },
+    graphql: {
+      outputType: "JSON",
+      inputType: "JSON",
+      canRead: ["admins"],
+      canCreate: ["members"],
+      validation: {
+        optional: true,
+        blackbox: true,
+      },
+    },
+    form: {
+        label: "Event Data",
+        // hidden: true
+    }
+  } satisfies NewCollectionFieldSpecification<"UltraFeedEvents">,
+
+  feedItemId: {
+    database: {
+      type: "TEXT", 
+      nullable: true, // TODO: once this is being provided, make it required
+    },
+    graphql: {
+      outputType: "String",
+      inputType: "String",  // TODO: once this is being provided, make it required
```

Comment:
Are these real TODOs?

Reply:
yes, either this field gets used or should be dropped